Avoid a context switch in the history api

bdraco opened the pull request:
## Proposed change
<!-- 
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
The history api was creating a job to fetch the
states and another job to convert the states to
json. This can be done in a single job which
decreases the overhead of the operation.

This should be 1:1 functionally equivalent to the existing code and already covered by the tests

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box! 
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [x] Code quality improvements to existing code or addition of tests

## Example entry for `configuration.yaml`:
<!--
  Supplying a configuration snippet, makes it easier for a maintainer to test
  your PR. Furthermore, for new integrations, it gives an impression of how
  the configuration would look like.
  Note: Remove this section if this PR does not have an example entry.
-->

```yaml
# Example configuration.yaml

```

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

elupus reviewed (commented):
I would have put the _get_sorted_significant_states_as_json() function as an internal function to the get() function. It will then have access to all variables without the need to pass them all as parameters. We are awaiting this to finish anyway.

bdraco committed:
Avoid a context switch in the history api

The history api was creating a job to fetch the
states and another job to convert the states to
json. This can be done in a single job which
decreases the overhead of the operation.

bdraco commented:
> I would have put the _get_sorted_significant_states_as_json() function as an internal function to the get() function. It will then have access to all variables without the need to pass them all as parameters. We are awaiting this to finish anyway.

Good call. Much smaller `git diff -w`

```
diff --git a/homeassistant/components/history/__init__.py b/homeassistant/components/history/__init__.py
index 6fc68b2833..0e7dc25072 100644
--- a/homeassistant/components/history/__init__.py
+++ b/homeassistant/components/history/__init__.py
@@ -335,8 +335,10 @@ class HistoryPeriodView(HomeAssistantView):
 
         hass = request.app["hass"]
 
-        result = await hass.async_add_job(
-            get_significant_states,
+        def _sorted_significant_states_json():
+            """Fetch significant stats from the database as json."""
+
+            result = get_significant_states(
                 hass,
                 start_time,
                 end_time,
@@ -348,7 +350,9 @@ class HistoryPeriodView(HomeAssistantView):
             result = list(result.values())
             if _LOGGER.isEnabledFor(logging.DEBUG):
                 elapsed = time.perf_counter() - timer_start
-            _LOGGER.debug("Extracted %d states in %fs", sum(map(len, result)), elapsed)
+                _LOGGER.debug(
+                    "Extracted %d states in %fs", sum(map(len, result)), elapsed
+                )
 
             # Optionally reorder the result to respect the ordering given
             # by any entities explicitly included in the configuration.
@@ -363,7 +367,9 @@ class HistoryPeriodView(HomeAssistantView):
                 sorted_result.extend(result)
                 result = sorted_result
 
-        return await hass.async_add_job(self.json, result)
+            return self.json(result)
+
+        return await hass.async_add_executor_job(_sorted_significant_states_json)
 
 
 class Filters:
```

MartinHjelmare commented on homeassistant/components/history/__init__.py:
```diff
@@ -335,35 +335,41 @@ async def get(self, request, datetime=None):
-            result = sorted_result
-
-        return await hass.async_add_job(self.json, result)
+        def _sorted_significant_states_json():
```
Should we make this function a helper method so we don't need to define it every request?

elupus replied:
I don't think the overhead is anything to talk about in python. It's also makes it much larger with all parameters to pass in.

bdraco replied:
The internal function is easier to read, but it does perform as well (its close).   The history api gets pounded with requests if you have a lot of gauges so anything we can get out of it is a nice improvement.   I went back to the original solution as the variable passing isn't in another file and is right below so it doesn't make it significantly harder to maintain.

bdraco committed:
Revert to original solution to avoid function redefine each call

MartinHjelmare reviewed (approved):
Looks good!
